### ISSUE TITLE
External WGPU surfaces / BYOW

### ISSUE THREAD

**Issue by littledivy** [denoland/deno#21713]
Goals:

- De-facto windowing solution.
- BYOW: Bring your own window.
- Deno should **not link** to native windowing systems like X11.
- Deno should not have to be **forked** for windowing support.

```jsx
/* any windowing library */
import { WindowBuilder } from "https://deno.land/x/sdl2/mod.ts";
const win = new WindowBuilder("Hello WGPU").build();

/* [string, Deno.PointerValue, Deno.PointerValue] */
const [system, windowHandle, displayHandle] = window.rawHandle();

/* GPUCanvasContext */
const context = Deno.createWindowSurface(system, windowHandle, displayHandle);

context.configure({ });

for (const events of win.events()) {
  const texture = context.getCurrentTexture();
  texture.present();
}
```

Proposed: Add a `Deno.createWindowSurface` API:

1. Creates a `raw_window_handle::RawWindowHandle` implementation from `windowHandle` and `displayHandle`
2. Internally calls `instance->create_surface(&win)`
3. Surface is returned wrapped in as a GPUCanvasContext (@crowlKats has already implemented presentation APIs)

```tsx
function createWindowSurface(
  system: "win32" | "winrt" | "cocoa" | "x11" | "wayland",
  winHandle: Deno.PointerValue | null,
  displayHandle: Deno.PointerValue | null
): GPUCanvasContext;
```

Pros: 

- Single binary. No extensions or Deno forks.
- Does not link CLI to windowing systems.
- Plug and play FFI windowing libraries with Deno WebGPU.
- `deno compile`-able.

Cons: 

- Very low level
- `displayHandle` and `winHandle` can be invalid handles. Creating them requires `--allow-ffi --unstable`

I have a working implementation and a few demos I will share shortly.

cc-ing for visibility: @DjDeveloperr @load1n9 @eliassjogreen

**Comment by littledivy** [comment]
Some demos using [deno_sdl2](https://github.com/littledivy/deno_sdl2):

1. Shaderplay, a WGSL live playground: https://github.com/littledivy/shaderplay

2. Conway's game of life (compute + render pass example): https://github.com/littledivy/deno_sdl2/blob/webgpu_external_surface/webgpu-examples/gameoflife.ts

4. Orillusion, a WebGPU game engine example scene: https://github.com/littledivy/deno_sdl2/blob/webgpu_external_surface/webgpu-examples/orillusion.ts

4. [`brendan-duncan/gpucraft`](https://github.com/brendan-duncan/gpucraft/tree/main): Minecraft-like game using WebGPU. 
Gameplay in Deno: https://youtu.be/vTJvjBG-GLs

**Comment by lino-levan** [comment]
This makes a ton of sense. Massive +1 from my side. 

> Deno should not have to be forked for windowing support.

This is definitely the most important point for me. Maintaining a fork just doesn't make sense in the long run.

**Comment by crowlKats** [comment]
> This is definitely the most important point for me. Maintaining a fork just doesn't make sense in the long run.

just as an FYI: Forking hasnt been necessary for a long time, and could be achieved with the deno crates.

**Comment by abflow** [comment]
Is it something that could be paired with Skia, or CanvasKit with WebAssembly ?
https://skia.org/docs/user/modules/canvaskit/

**Comment by crowlKats** [comment]
No, since skia is it own rendering system, which is unrelated to WebGPU

**Comment by abflow** [comment]
Thanks, I thought that as skia can render graphics to WebGL (at least concerning the WebAssembly CanvasKit module), it could have rendered to WebGPU as well

**Comment by littledivy** [comment]
Updated the proposal. Renamed API to `Deno.UnsafeWindowSurface`
